### PR TITLE
support prefix scans in ResponsiveKeyValueStore

### LINKED ISSUE #408: Support prefixScan on ResponsiveKeyValueStore
Today we support `range` on `ResponsiveKeyValueStore` but we neglected to implement `prefixScan` (which can be implemented similarly except setting to/from to be one byte incremented).

### PR DESCRIPTION
**Pull request by agavra**
fixes #408 

Support prefix scans on:
- `CassandraKeyValueTable`
- `MongoKeyValueTable`
- `InMemoryKeyValueTable`

We don't support range scans on Global tables (yet), `CassandraFactTable` or `RS3KeyValueTable` yet.

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java [hunk 1]
```diff
@@ -383,6 +402,27 @@ public KeyValueIterator<Bytes, byte[]> range(
...
+  @Override
+  public <PS extends Serializer<P>, P> KeyValueIterator<Bytes, byte[]> prefix(
+      final P prefix,
+      final PS prefixKeySerializer,
+      final int kafkaPartition,
+      final long streamTimeMs
+  ) {
+    final Bytes from = Bytes.wrap(prefixKeySerializer.serialize(null, prefix));
+    final Bytes to = Utils.incrementWithoutOverflow(from);
+    return doRange(kafkaPartition, from, to, streamTimeMs, this.prefix);
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java:417]: very nitpicky but why use `range` above but `this.prefix` here?  Can only afford one `this` ? 😛

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java [hunk 2]
```diff
@@ -383,6 +402,27 @@ public KeyValueIterator<Bytes, byte[]> range(
...
+  }
+
+  private KeyValueIterator<Bytes, byte[]> doRange(
+      final int kafkaPartition,
+      final Bytes from,
+      final Bytes to,
+      final long streamTimeMs,
+      final PreparedStatement rangeStatement) {
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java:425]:
```suggestion
      final PreparedStatement rangeStatement
    ) {
```

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/CassandraKVTableIntegrationTest.java [hunk 3]
```diff
@@ -196,6 +197,40 @@ public void shouldReturnRangeKeysInLexicalOrderAcrossMultipleSubPartitions() {
     }
   }
 
+  @Test
```
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/kafka/internal/db/CassandraKVTableIntegrationTest.java:200]: Can we add a test that hits the overflow case? Would be good to have one for each table type to make sure the underlying engine does the right thing

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java (hunk 1 above)
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java:417]: 😆 it's because this has `prefix` in the parameters so there's a scoping conflict whereas range doesn't have that. I'll use `this` for both for symmetry if you demand it haha

File: kafka-client/src/main/java/dev/responsive/kafka/internal/utils/Utils.java [hunk 4]
```diff
@@ -117,4 +118,12 @@ public static String extractThreadIdFromThreadName(final String threadName) {
     LOG.warn("Unable to parse the stream thread id, falling back to thread name {}", threadName);
     return threadName;
   }
+
+  public static Bytes incrementWithoutOverflow(final Bytes input) {
+    try {
+      return Bytes.increment(input);
+    } catch (final IndexOutOfBoundsException e) {
+      return null;
```
[the remark this answers is not in the document]
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/utils/Utils.java:126]: Good call, fixed bugs and added tests.

[the remark this answers is not in the document]
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/utils/Utils.java:126]: I fixed it for Mongo/In Memory but it's a bit more work for C* so I'll just create a ticket for that now.
